### DISCUSSION TITLE
Sending a large thumbnail URL in channel messages

### DISCUSSION THREAD

**Discussion by rudyzeinoun** [tulir/whatsmeow#498] [WhatsApp protocol Q&A] [answered]
I am sending an ExtendedTextMessage containing a link to a channel (newsletter) with the following properties:

```
ext := &waProto.ExtendedTextMessage{
		Text:                  proto.String(message.Message),
		CanonicalUrl:          proto.String(canonical),
		Title:                 proto.String(title),
		MatchedText:           proto.String(message.Url),
		Description:           proto.String(desc),
		ThumbnailWidth:        proto.Uint32(800),
		ThumbnailHeight:       proto.Uint32(384),
		InviteLinkGroupTypeV2: waProto.ExtendedTextMessage_DEFAULT.Enum(),
		PreviewType:           waProto.ExtendedTextMessage_NONE.Enum(),
		JpegThumbnail: jpegThumbnailBytes,
}
```

The received message on the channel has a small thumbnail preview.
I tried sending the same message from WhatsApp Web. The preview is a large full-width image. Using the event handler, I can see the same fields in the received message, with the exception of thumbnailDirectPath and thumbnailSha256 being missing in whatsmeow (not WhatsApp Web).

`21:33:46.091 [Main INFO] Received message 3EB088444F9ABD168211AC from 120363195752054779@newsletter (pushname: , timestamp: 2023-11-21 15:49:03 +0100 CET, type: text): extendedTextMessage:{text:"[truncated]"  matchedText:"[truncated]"  canonicalUrl:"[truncated]"  description:[truncated]"  previewType:NONE  jpegThumbnail:"[truncated]"  thumbnailDirectPath:"[truncated]"  thumbnailHeight:384  thumbnailWidth:800  inviteLinkGroupTypeV2:DEFAULT}  messageContextInfo:{messageSecret:"[truncated] "}
`

I saved the thumbnail file from the message with a large preview, and it is a rather very small JPEG (100x48).

I suspect there's a missing larger image file that's being sent with the message and not showing after parsing. Maybe we should attach a larger image that I'm not aware of?

Appreciate any help :) Thanks

**Comment by arugaz** [comment]
have you tried using cli.Upload for thumbnail?

```go
resp, err := cli.UploadNewsletter(context.Background(), yourthumb, whatsmeow.whatsmeow.MediaLinkThumbnail)
ext := &waProto.ExtendedTextMessage{
	...
	ThumbnailDirectPath: &resp.DirectPath,
	ThumbnailSha256:     resp.FileSHA256,
	ThumbnailEncSha256:  resp.FileEncSHA256,
}
```

**Comment by rudyzeinoun** [comment]
@arugaz thanks for your swift answer.
Yes, I tried your suggestion too. It makes no difference. I also tried either keeping JpegThumbnail or removing it.
When I remove JpegThumbnail and keep the other fields (ThumbnailDirectPath, ThumbnailSha256, ThumbnailEncSha256), the image disappears.

You can send from your phone a link on a channel, and you'll see the large preview. Printing out the message to the console doesn't show any difference. I tried using debugging mode, but the message contents are hard to parse as they're binary or encrypted?

Could it be a new field in the protobuf schema that isn't added?

Thanks.

```
ext := &waProto.ExtendedTextMessage{
    Text:                  proto.String(message.Message),
    CanonicalUrl:          proto.String(canonical),
    Title:                 proto.String(title),
    MatchedText:           proto.String(message.Url),
    Description:           proto.String(desc),
    ThumbnailWidth:        proto.Uint32(800),
    ThumbnailHeight:       proto.Uint32(384),
    InviteLinkGroupTypeV2: waProto.ExtendedTextMessage_DEFAULT.Enum(),
    PreviewType:           waProto.ExtendedTextMessage_NONE.Enum(),
}

if image_url != "" {

    // Download image and send as bytes
    image_resp, image_err := http.Get(image_url)
    log.Infof("Download image %s", image_url)
    if image_err == nil {
	    if image_resp.StatusCode == 200 && image_resp.ContentLength > 0 {
		    log.Infof("Image size %i", image_resp.ContentLength)
    
		    image_body, err := ioutil.ReadAll(image_resp.Body)
		    if err != nil {
			    response["error"] = true
			    response["message"] = "Cannot read image"
			    response["details"] = err
			    json.NewEncoder(w).Encode(response)
			    return
		    }
    
		    image_obj, _, err := image.Decode(bytes.NewReader(image_body))
		    if err != nil {
			    response["error"] = true
			    response["message"] = err.Error()
			    json.NewEncoder(w).Encode(response)
			    return
		    }
    
		    resp, _ := cli.Upload(context.Background(), image_body, whatsmeow.MediaImage)
		    ext.ThumbnailDirectPath = &resp.DirectPath
		    ext.ThumbnailSha256 = resp.FileSHA256
		    ext.ThumbnailEncSha256 = resp.FileEncSHA256
    
		    newImage := resize.Resize(800, 384, image_obj, resize.Lanczos3)
		    newbuffer := new(bytes.Buffer)
		    _ = jpeg.Encode(newbuffer, newImage, nil)
		    ext.JpegThumbnail = newbuffer.Bytes()
	    }
    }
    defer link_res.Body.Close()

}

msg = &waProto.Message{ExtendedTextMessage: ext}
```

**Reply by arugaz** [reply]
Can you try this

```go
image_obj, _, err := image.Decode(bytes.NewReader(image_body))
if err != nil {
  response["error"] = true
  response["message"] = err.Error()
  json.NewEncoder(w).Encode(response)
  return
}

newImage := resize.Resize(800, 384, image_obj, resize.Lanczos3)
newbuffer := new(bytes.Buffer)
_ = jpeg.Encode(newbuffer, newImage, nil)
ext.JpegThumbnail = newbuffer.Bytes()

resp, _ := cli.UploadNewsletter(context.Background(), newbuffer.Bytes(), whatsmeow.MediaLinkThumbnail)
ext.ThumbnailDirectPath = &resp.DirectPath
ext.ThumbnailSha256 = resp.FileSHA256
ext.ThumbnailEncSha256 = resp.FileEncSHA256
```

**Comment by rudyzeinoun** [comment]
@arugaz I tried your code, and it gives the same result (small image thumb).

I do notice one thing that might be helpful. I sent the link from my phone, and I save the JPEG thumbnail received via whatsmeow. It is very tiny 100x48. The 800x384 image is nowhere to be found in the received message on whatsmeow. Although the fields ThumbnailWidth and ThumbnailHeight are equal to 800 and 384.

**Reply by arugaz** [reply]
yea, i just found out that you have to [resize](https://github.com/WhiskeySockets/Baileys/blob/e24947e55ad6e3370de867b258bb33bb9e828c56/src/Utils/messages-media.ts#L93) the image as small as possible, but you can set the thumb width and height with the default image size

**Reply by arugaz** [reply]
have you tried resize the image to 100x48 in your code, does it work?

**Reply by rudyzeinoun** [reply]
Yes I tried it with a small thumbnail. It simply appears on the device as a small thumbnail but pixelated.
Nothing else is telling WhatsApp to use a larger image format.

Worth nothing that I am setting
`previewType:NONE`
because that's the preview type logged for the large preview by whatsmeow.

Also, here is the debug log from a large preview message:
` <message from="120363195752054779@newsletter" id="9A64FF8AD60F3C0B5F92FF88CCD82273" is_sender="true" server_id="135" t="1700649153" type="media"><plaintext mediatype="url"></plaintext></message>
`
Does it provide any insight on the message type to be sent?

**Comment by arugaz** [comment, marked as answer]
theres other func for upload media to channel

https://github.com/tulir/whatsmeow/blob/9eb41978f180d5a37577253406ad6315166bd67a/upload.go#L124-L130

**Reply by rudyzeinoun** [reply]
This works! Thanks a lot. If only I had read the documentation more carefully.

For others doing this, one additional change is also required to complete this:

```
resp, err := cli.SendMessage(context.Background(), recipient, msg, whatsmeow.SendRequestExtra{
	MediaHandle: resp.Handle,
})
```
